### PR TITLE
Add stub `specs-model` tool

### PR DESCRIPTION
**Pull request by konrad-jamrozik**
Contributes to:

- https://github.com/Azure/azure-sdk-tools/issues/8203

The added CI pipeline runs:
- https://github.com/Azure/azure-rest-api-specs/actions/workflows/specs-model-test.yaml

@mikeharder @weshaggard I left bunch of "FYI" comments explaining some nontrivial decisions I made.

### CODE REVIEW

File: .github/workflows/specs-model-test.yaml
Comment (konrad-jamrozik) [inline .github/workflows/specs-model-test.yaml:8]: FYI The line `- users/kojamroz/specs_model_1` to be removed before merging. it is here for testing.

File: eng/.vscode/settings.json
Comment (konrad-jamrozik) [inline eng/.vscode/settings.json:1]: FYI I very much prefer to open `eng` directory in vscode, as if I open entire repo it takes forever for `vscode` search to index all files, and so on.

However, I am unsure how to approach avoiding deduping settings. Looks like VSCode does not support "includes":

- https://github.com/microsoft/vscode/issues/15909

**Comment by konrad-jamrozik** [comment]
FYI I am unsure how to avoid the npm errors in `npm ls -a`:

https://github.com/Azure/azure-rest-api-specs/actions/runs/9935390601/job/27441478831#step:5:5

They started happening ([first build with them](https://github.com/Azure/azure-rest-api-specs/actions/runs/9935030411)) once I did `npm install` in specs repo root dir due to [this error](https://github.com/Azure/azure-rest-api-specs/actions/runs/9935009655/job/27440383951#step:4:7):

> npm error `npm ci` can only install packages when your package.json and package-lock.json or npm-shrinkwrap.json are in sync. Please update your lock file with `npm install` before continuing.

File: eng/tools/specs-model/eslint.config.js
Comment (konrad-jamrozik) [inline eng/tools/specs-model/eslint.config.js:1]: FYI it took me almost 3 hours to get this eslint work as intended, with no false positive typing errors, etc.

My take is to start with this very strict ruleset and gradually adjust the rules in the `rules: {` section, providing in comments our explanation why we adjusted some rules. This way we will gradually build explicit code conventions, immediately backed by automation.

File: eng/tools/suppressions/package-lock.json
Comment (konrad-jamrozik) [inline eng/tools/suppressions/package-lock.json:1]: FYI I am unsure why this `package-lock.json` was missing. I added it because I wanted to `npm run test` from the `suppressions` directory.

File: .github/workflows/specs-model-test.yaml
Comment (konrad-jamrozik) [inline .github/workflows/specs-model-test.yaml:26]: FYI while I added mandatory eslint and prettier checks to specs-model, I would be in favor in enabling it for everything in `eng`. I am happy to help adjusting the existing code and linting rules to make eslint and prettier pass on everything.

File: eng/tools/specs-model/package.json
Comment (konrad-jamrozik) [inline eng/tools/specs-model/package.json:40]: FYI this is necessary because `typescript-eslint` has problems with `eslint 9+` during `npm install`.

File: eng/tools/suppressions/src/suppressions.ts
Comment (konrad-jamrozik) [inline eng/tools/suppressions/src/suppressions.ts:40]: FYI changes in this file are caused by prettier

File: eng/tools/specs-model/package.json
Comment (konrad-jamrozik) [inline eng/tools/specs-model/package.json:35]: FYI some of the dependencies I use here are newest, newer than in `suppressions`. Like `typescript` or `vitest`. Not sure what is our approach for keeping them in sync.

File: eng/.vscode/settings.json
Comment (mikeharder) [reply eng/.vscode/settings.json:1]: I open repo root in VSCode and perf seems fine, but I don't mind duplicating a few settings.

File: .github/workflows/specs-model-test.yaml
Comment (mikeharder) [reply .github/workflows/specs-model-test.yaml:8]: FYI, you should be able to test a new Action by just opening a fork pull request.  Unlike DevOps pipelines, you don't need to create a branch in the main fork.

File: eng/tools/specs-model/.prettierignore
Comment (mikeharder) [inline eng/tools/specs-model/.prettierignore:5]:
```suggestion
```

I don't think this is needed

File: eng/.vscode/settings.json
Comment (mikeharder) [reply eng/.vscode/settings.json:1]: Does opening the eng folder work with the root package-lock.json?  Because we only want a single package-lock.json in repo root, not a lockfile per folder.

File: eng/tools/specs-model/.prettierrc.mjs
Comment (mikeharder) [inline eng/tools/specs-model/.prettierrc.mjs:7]:
```suggestion
  semi: true,
```

I think we should align with the format used in the typespec repo as much as possible:

https://github.com/microsoft/typespec/blob/main/.prettierrc.json

In particular, I feel strongly that semicolons should be required.

I also think we should use a single format for all TS code under eng.  After this PR is merged, I can update all the other projects to use the same settings.

File: eng/tools/specs-model/package-lock.json
Comment (mikeharder) [inline eng/tools/specs-model/package-lock.json:1]: Repo should only have one package-lock.json in the repo root.

File: eng/tools/specs-model/package.json
Comment (mikeharder) [reply eng/tools/specs-model/package.json:35]: As much as possible, I have been aligning versions in `azure-rest-api-specs` with the `microsoft/typespec` repo.  Additionally, all tools under eng should be aligned.

File: eng/tools/specs-model/package.json
Comment (mikeharder) [inline eng/tools/specs-model/package.json:33]:
```suggestion
    "eslint": "^8.57.0",
```

I think we should stay on `eslint` v8 until `typescript-eslint` supports v9.  This also aligns with `microsoft/typespec`.

File: eng/tools/specs-model/package.json
Comment (mikeharder) [reply eng/tools/specs-model/package.json:40]: I think we should use `eslint@^8` instead.

File: eng/tools/specs-model/src/index.ts
Comment (mikeharder) [inline eng/tools/specs-model/src/index.ts:32]:
```suggestion
export { getSpecsModel }
```

File: eng/tools/specs-model/tsconfig.json
Comment (mikeharder) [inline eng/tools/specs-model/tsconfig.json:13]:
```suggestion
    // We don't need it anyway, as all sources are in .ts except the 3-line cmd entry-point.
```

File: eng/tools/suppressions/package-lock.json
Comment (mikeharder) [inline eng/tools/suppressions/package-lock.json:1]: Delete

File: .github/workflows/_reusable-eng-tools-test.yaml
Comment (mikeharder) [inline .github/workflows/_reusable-eng-tools-test.yaml:11]: Fine for now, but after this PR, I think we should enable lint and prettier consistently for all packages under eng.

File: eng/tools/specs-model/eslint.config.js
Comment (mikeharder) [reply eng/tools/specs-model/eslint.config.js:1]: I'm fine using whatever rules you want as a starting point, and we can discuss specific points of disagreement.

**Review by mikeharder — CHANGES_REQUESTED** [review]
Feedback in comments

**Comment by mikeharder** [comment]
> FYI I am unsure how to avoid the npm errors in `npm ls -a`:
> 
> https://github.com/Azure/azure-rest-api-specs/actions/runs/9935390601/job/27441478831#step:5:5
> 
> They started happening ([first build with them](https://github.com/Azure/azure-rest-api-specs/actions/runs/9935030411)) once I did `npm install` in specs repo root dir due to [this error](https://github.com/Azure/azure-rest-api-specs/actions/runs/9935009655/job/27440383951#step:4:7):
> 
> > npm error `npm ci` can only install packages when your package.json and package-lock.json or npm-shrinkwrap.json are in sync. Please update your lock file with `npm install` before continuing.

If `npm ls -a` is reporting an error, I think it usually means the lockfile needs to be updated again using `npm install`, or possibly `npm update` (though this updates all dependencies to latest, so we try to do it sparingly, not in every PR).

File: eng/tools/suppressions/package-lock.json
Comment (mikeharder) [reply eng/tools/suppressions/package-lock.json:1]: The standard workflow in `azure-rest-api-specs` is to run `npm ci` in the repo root to install deps for all folders in the repo.  This should allow you to exec `npm run test` under any folder under `eng`.

It's fine if you want to work locally on a subset of the repo, but I'd like to keep the repo itself as clean as possible (so only a single root package-lock.json).

File: eng/tools/specs-model/.prettierrc.mjs
Comment (konrad-jamrozik) [reply eng/tools/specs-model/.prettierrc.mjs:7]: Totally, I'll copy over typespec prettier. I am just curious why you prefer semicolons so much ;)

Comment (mikeharder) [reply eng/tools/specs-model/.prettierrc.mjs:7]: > Totally, I'll copy over typespec prettier. I am just curious, why you prefer semicolons so much?

Consistency with languages that require them (C, C#, Java).

File: eng/tools/specs-model/package.json
Comment (konrad-jamrozik) [reply eng/tools/specs-model/package.json]: I aligned versions in `package.json` files for specs-model and suppressions, by downgrading specs-model versions to match.

I also reordered the keys a bit, to be in following order:

name
private
main
bin
scripts
engines
dependencies
devDependencies

The reasoning here is that `bin` and `scripts` should be near the top, as these are the entry-points.

File: eng/tools/specs-model/package-lock.json
Comment (konrad-jamrozik) [reply eng/tools/specs-model/package-lock.json:1]: I also added entry to `.gitignore` and `eng/README.md` explaining this.

File: eng/tools/specs-model/package.json
Comment (konrad-jamrozik) [reply eng/tools/specs-model/package.json]: I used `"@types/node": "^18.19.31"` from suppressions package.json

File: eng/tools/specs-model/src/index.ts
Comment (konrad-jamrozik) [reply eng/tools/specs-model/src/index.ts]: Fixed and added following entry to eslint config:

``` jsonc
      // We prefer to have explicitly import at the top of the file, even if the same element is exported again,
      // which we do in index.ts files.
      // https://github.com/sindresorhus/eslint-plugin-unicorn/blob/main/docs/rules/prefer-export-from.md
      "unicorn/prefer-export-from": ["error", { ignoreUsedVariables: true }],
```

File: eng/README.md
Comment (konrad-jamrozik) [inline eng/README.md:1]: FYI @mikeharder I added this to capture the conventions you mentioned on this PR review.

File: eng/README.md
Comment (konrad-jamrozik) [inline eng/README.md:18]: FYI this is my proposal for the key order. I moved `bin` and `scripts` close to top as package entry-points people probably want to refer to.

File: .github/workflows/specs-model-test.yaml
Comment (konrad-jamrozik) [reply .github/workflows/specs-model-test.yaml]: I am unsure how to do that. I do hope I won't have to fiddle with forks to test my action. I added the ability to manually trigger the workflow.

**Comment by konrad-jamrozik** [comment]
@mikeharder I copied over prettier config, but I am postponing using the import reorder plugin due to:

> Checking formatting...
> Error:  Cannot find module '/home/runner/work/azure-rest-api-specs/azure-rest-api-specs/eng/tools/specs-model/node_modules/prettier-plugin-organize-imports/index.js' imported from /home/runner/work/azure-rest-api-specs/azure-rest-api-specs/node_modules/prettier/index.mjs
> Error: Process completed with exit code 1.

https://github.com/Azure/azure-rest-api-specs/actions/runs/9946326008/job/27476558336?pr=29820#step:8:10

File: eng/.vscode/settings.json
Comment (konrad-jamrozik) [reply eng/.vscode/settings.json:1]: I think it does, but not sure how I would now it does not work. In any case, I suggest we figure out the .vscode settings situation in a follow-up PR.

File: eng/tools/specs-model/package.json
Comment (mikeharder) [inline eng/tools/specs-model/package.json:24]: Where is lodash used in your code?  I didn't see it.

File: eng/tools/specs-model/src/index.ts
Comment (mikeharder) [inline eng/tools/specs-model/src/index.ts:19]: Why the trailing underscore on `arguments_`?  Is this a coding convention to signify something?

Comment (konrad-jamrozik) [reply eng/tools/specs-model/src/index.ts:19]: unicorn linter forbade using `args` (no abbreviations) and arguments appears to be reserved keyword. I am happy to disable this rule for `args` and use `args`.

File: package-lock.json
Comment (mikeharder) [inline package-lock.json:1]: I think you might need to regenerate package-lock.json.  For example, I still see `vitest@2.0.2` but I think you downgraded in your pacakge.json.

**Review by mikeharder — CHANGES_REQUESTED** [review]
Feedback in comments

File: eng/tools/specs-model/src/index.ts
Comment (konrad-jamrozik) [reply eng/tools/specs-model/src/index.ts]: Fixed with linter config update:

```
      // We allow some abbreviations that we like.
      // https://github.com/sindresorhus/eslint-plugin-unicorn/blob/main/docs/rules/prevent-abbreviations.md
      "unicorn/prevent-abbreviations": {
        allowList: {
          args: true,
        },
      },
```

Comment (mikeharder) [reply eng/tools/specs-model/src/index.ts]: I agree in principle, but I'm not sure how "prevent abbreviations" will work in practice.  I think some abbreviations are good (e.g. `args`, `dir`, etc), while others are bad (e.g. `cust` instead of `customer`).  It's more a matter of taste than "anything not in the dictionary should be banned".

But I'm willing to give it a try, so I'm fine with disabling the rule for `args` for now.

File: package-lock.json
Comment (konrad-jamrozik) [reply package-lock.json:1]: Ah sorry, I didn't realize I need to redo package-lock.json to get rid of unused deps, and `npm install` does not do it.

File: eng/tools/specs-model/src/index.ts
Comment (konrad-jamrozik) [reply eng/tools/specs-model/src/index.ts]: Yeah, I have similar attitude - if it turns out this rule is too against our preferences, we can outright disable it. But let's give it a try.

File: eng/tools/specs-model/package.json
Comment (konrad-jamrozik) [reply eng/tools/specs-model/package.json:24]: Not yet used, but I am quite sure I will use it. But I can remove it for now.

Comment (mikeharder) [reply eng/tools/specs-model/package.json]: If eslint (or any similar tools) could enforce "every dependency must actually be used", I'd support enabling the check.

File: package-lock.json
Comment (mikeharder) [inline package-lock.json:43]: Looks like a bad update to package-lock.json?

**Review by mikeharder — APPROVED** [review]
Approved pending fix to package-lock.json

File: package-lock.json
Comment (konrad-jamrozik) [reply package-lock.json:43]: Unsure what happened here. I am also running into this now:

https://github.com/Azure/azure-rest-api-specs/actions/runs/9947828905/job/27481266070?pr=29820#step:4:12

```
Error: npm error tsconfig.json(5,5): error TS6053: File '@tsconfig/strictest/tsconfig.json' not found.
npm error A complete log of this run can be found in: /home/runner/.npm/_logs/2024-07-15T22_59_16_546Z-debug-0.log
Error: Process completed with exit code 1.
```

It did work before, so I am yet unsure what's up.

Comment (mikeharder) [reply package-lock.json]: > Unsure what happened here. I am also running into this now:
> 
> https://github.com/Azure/azure-rest-api-specs/actions/runs/9947828905/job/27481266070?pr=29820#step:4:12
> 
> ```
> Error: npm error tsconfig.json(5,5): error TS6053: File '@tsconfig/strictest/tsconfig.json' not found.
> npm error A complete log of this run can be found in: /home/runner/.npm/_logs/2024-07-15T22_59_16_546Z-debug-0.log
> Error: Process completed with exit code 1.
> ```
> 
> It did work before, so I am yet unsure what's up.

I pushed changes to fix `package-lock.json` in this PR.  How I did it:

1. `git clean -xdf`
  - remove all cached files in your local repo folder
2.  `cp ../specs-main/package-lock.json ./`
  - Overwrite `package-lock.json` in local folder with copy from main
3. `npm install`
  - Should make minimal changes to `package-lock.json`, to support whatever dep changes you make in `package.json`.

File: eng/README.md
Comment (konrad-jamrozik) [inline eng/README.md:24]: @mikeharder can you review this guidance?

File: eng/README.md
Comment (mikeharder) [inline eng/README.md:35]: - To avoid conflicting changes when updating `package-lock.json`, use the latest LTS version of Node and the bundled version of npm, but no newer.  Commands `nvm install --lts` (Linux) and `nvm install lts` (Windows) should install the correct versions.

Comment (mikeharder) [inline eng/README.md:35]:
```suggestion
- To avoid inconsistent content of `package-lock.json` due to bugs like [npm/cli #7384],
  we ensure to use the same node and npm version to update `package-lock.json`.  
  As of 7/15/2024 this is `node 20.15.1` and `npm 10.7.0`.
```

**Review by mikeharder — APPROVED** [review]

File: eng/tools/specs-model/.prettierignore
Comment (weshaggard) [inline eng/tools/specs-model/.prettierignore:1]: Should these config files be moved up into the tools directory so that we have a shared set for all our ts tools?

Comment (mikeharder) [reply eng/tools/specs-model/.prettierignore:1]: Eventually, yes.  But we will need to migrate each tool individually.

**Pull request merged**
